rfc: construct tensor without user handle (#1739)
oneapi-src/oneDNN: rfcs/20230927-construct-tensor-without-user-handle/README.md

# RFC: construct tensor without user handle

## Introduction & Motivation

> In benchdnn and unit tests, users often need to allocate memory themselves. 

Reviewer: You may add the current status of the API here, to make the introduction more clear.

Incorrect allocation methods may cause some unexpected errors, such as memory 

> release errors caused by `test::vector`. In fact, the size of the memory can be 

Reviewer: Do you want to elaborate what is memory release errors caused by `test::vector` here?

> release errors caused by `test::vector`. In fact, the size of the memory can be 
> obtained based on the logical tensor and layout. This RFC mainly implements the 
> ability to create tensors without user handle.

Reviewer: ```suggestion
release errors caused by `test::vector`. In fact, in scenarios like unit test, the library can take the control of allocating tensor by itself, since the size of the tensor can be 
obtained based on the logical tensor and layout. This RFC mainly implements the 
ability to create tensors without user handle.
```

///       allocate the buffer for the tensor. In this case the library
///       owns the buffer.
///     - DNNL_MEMORY_NONE to create tensor without an underlying buffer.
tensor(const logical_tensor &lt, const engine &aengine, void *handle) {
    dnnl_graph_tensor_t t = nullptr;
    error::wrap_c_api(
            dnnl_graph_tensor_create(&t, &(lt.data), aengine.get(), handle),
            "could not create tensor object with the logical_tensor, "
            "engine, and handle");
    reset(t);
}
```

### Example

Below is pseudo-code that demonstrates the use of the C and C++ API.
```c
// Create a engine
dnnl_engine_t engine;
dnnl_engine_create(&engine, dnnl_cpu, 0);
// Create a logical tensor
dnnl_graph_logical_tensor_t lt;
const size_t id = 0;
dnnl_dims_t dims = {1, 2, 3, 4};
// Init logical tensor
dnnl_graph_logical_tensor_init_with_dims(&lt, id, dnnl_f32, 4, dims, 
    dnnl_graph_layout_type_strided, dnnl_graph_tensor_property_undef);
// Create a tensor
dnnl_graph_tensor_t tensor;
dnnl_graph_tensor_create(&tensor, &lt, engine, DNNL_MEMORY_ALLOCATE);
// Destroy the tensor
dnnl_graph_tensor_destroy(tensor);
// Destroy the engine
dnnl_engine_destroy(engine);
```

```c++

> // Create a engine

Reviewer: ```suggestion
// Create an engine
```

# RFC for Graph API: construct tensor without user handle

## Introduction & Motivation

Currently, in Graph API level, users need to manually allocate memory and manage 
its lifecycle before constructing a tensor. Incorrect allocation and deallocation 
methods may cause some unexpected errors, such as memory release errors caused 

> by `test::vector`(use one context allocate memory and another context to deallocate 

Reviewer: ```suggestion
by `test::vector`(use one context to allocate memory and another context to deallocate 
```

it). In fact, in scenarios like unit test, the library can take the control of 
allocating and deallocating memory by itself, since the size of the memory can be 
obtained based on the logical tensor and layout, memory allocation can be performed 
accordingly when a tensor is constructed. When a tensor is destroyed, the memory is 
deallocated. In order to make library more user-friendly, this RFC mainly implements 
the ability to construct tensor without user handle.

## Proposal

### Option 1

Referring to the `dnnl_memory_create` API in the primitive, reuse the already 
used `dnnl_graph_tensor_create` API. When `handle` parameter is set to 
`DNNL_MEMORY_ALLOCATE`, the library will perform memory allocation and 
automatically release the memory when the current tensor is destroyed. When the 

> handle parameter is a regular pointer, the library will only set the pointer.

Reviewer: Also need to mention DNNL_MEMORY_NONE here? Since it's also newly added

# RFC: construct tensor without user handle

## Introduction & Motivation

> In benchdnn and unit tests, users often need to allocate memory themselves. Incorrect allocation methods may cause some unexpected errors, such as memory release errors caused by `test::vector`. In fact, the size of the memory can be obtained based on the logical tensor and layout. This RFC mainly implements the ability to create tensors without user handle.

Reviewer 2: You may need to change the format to make sure each line is less than 80 characters. It helps for review~
